[FEATURE] ZEP - PG `SqlYearMonthSplitter` serialization + deserialization

Kilo59 opened the pull request:
## Changes proposed in this pull request:
- Updated the `TableAsset.column_splitter` field to only accept concrete splitters
  - Update the `SqlYearMonthSplitter` to only accept the literal string `'split_on_year_and_month'` for `method_name`
  -  Update the `SqlYearMonthSplitter` to only accept the literal strings `'year'` or `'month'` for `param_names`
  - Updated base `ColumnSplitter` class to only accept a `method_name` that is a member of the `data_splitter.py` `SplitterMethod` enum.
- Added tests that specifically ensure config errors are raised

## References
https://docs.pydantic.dev/usage/types/#literal-type

### TODO

- [x] merge #6580 
- [x] add some tests
- [x] raise error if `method_name` is not a member of the `execution_engine.split_and_sample.data_splitter.SplitterMethod` enum https://github.com/great-expectations/great_expectations/blob/c0aa33feedb23671d75f3f15408171c8ee417c3b/great_expectations/execution_engine/split_and_sample/data_splitter.py#L47-L60

### Definition of Done

- [x] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/docs/contributing/style_guides/code_style)
- [x] I have performed a [self-review](https://docs.greatexpectations.io/docs/contributing/contributing_checklist) of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added [unit tests](https://docs.greatexpectations.io/docs/contributing/contributing_test#writing-unit-and-integration-tests) where applicable and made sure that new and existing tests are passing.
- [x] I have run any local integration tests and made sure that nothing is broken.

Kilo59 committed:
add variable tuple types

Kilo59 committed:
ignore some table_metrics issues

Kilo59 committed:
Merge remote-tracking branch 'origin/develop' into develop

Kilo59 committed:
Merge remote-tracking branch 'origin/develop' into develop

Kilo59 committed:
Merge remote-tracking branch 'origin/develop' into develop

Kilo59 committed:
Merge remote-tracking branch 'origin/develop' into develop

Kilo59 committed:
Merge remote-tracking branch 'origin/develop' into develop

Kilo59 committed:
Merge remote-tracking branch 'origin/develop' into develop

Kilo59 committed:
missing comma `,` in requirements

Kilo59 committed:
Merge remote-tracking branch 'origin/develop' into develop

Kilo59 committed:
Merge remote-tracking branch 'origin/develop' into develop

Kilo59 committed:
Merge remote-tracking branch 'origin/develop' into develop

Kilo59 committed:
Merge remote-tracking branch 'origin/develop' into develop

Kilo59 committed:
Merge remote-tracking branch 'origin/develop' into develop

Kilo59 committed:
Merge remote-tracking branch 'origin/develop' into develop

Kilo59 committed:
Merge remote-tracking branch 'origin/develop' into develop

Kilo59 committed:
make `orderby` a regular field with a custom validator

Kilo59 committed:
ignore empty list type error

Kilo59 committed:
print resulting config

Kilo59 committed:
order_by serialization test

Kilo59 committed:
Merge branch 'develop' into f/1424/splitter-sorter-sede

Kilo59 committed:
deserialize from sorter DSL strings `-year` `+year`

Kilo59 committed:
Merge remote-tracking branch 'origin/f/1424/splitter-sorter-sede' into f/1424/splitter-sorter-sede

Kilo59 committed:
Merge branch 'develop' into f/1424/splitter-sorter-sede

Kilo59 committed:
additional sorter tests

Kilo59 committed:
TDD test for future BatchSorter serialization work

Kilo59 committed:
better named test

Kilo59 committed:
deal with `KeyError` and remove print statement

Kilo59 committed:
Merge branch 'develop' into f/1424/splitter-sorter-sede

Kilo59 committed:
fix incorrect type-ignore explanation

Kilo59 committed:
rename`_parse_order_by_sorter` private

make private

Kilo59 committed:
`Literal` for `method_name` to allow pydantic to load correct subclass

Kilo59 committed:
Merge branch 'develop' into f/1448/splitters-sede

ghost commented:
<details>
<summary><b>👇 <a href="https://app.codesee.io/r/reviews?pr=6595&src=https%3A%2F%2Fgithub.com%2Fgreat-expectations%2Fgreat_expectations">Click on the image</a> for a new way to code review</b>
</summary>

  - Make big changes easier — review code in small groups of related files
  - Know where to start — see the whole change at a glance
  - Take a code tour — explore the change with an interactive tour
  - Make comments and review — all fully sync’ed with github

    [Try it now!](https://app.codesee.io/r/reviews?pr=6595&src=https%3A%2F%2Fgithub.com%2Fgreat-expectations%2Fgreat_expectations)
</details>

[![Review these changes using an interactive CodeSee Map](https://s3.us-east-2.amazonaws.com/maps.codesee.io/images/github/great-expectations/great_expectations/6595/fcf5d6a7/3a94f0bd7ffe524a7d8466a145360ec240c377af.svg)](https://app.codesee.io/r/reviews?pr=6595&src=https%3A%2F%2Fgithub.com%2Fgreat-expectations%2Fgreat_expectations)

#### Legend
![CodeSee Map Legend](https://app.codesee.io/Legend_v2.png)

Kilo59 committed:
Merge branch 'develop' into f/1448/splitters-sede

Kilo59 committed:
update config tests

Kilo59 committed:
update test config.yaml

Kilo59 committed:
`column_splitter` union does not work in current form.

specifically it does a poor job of catching errors as long as it satisfies the minimal generic `ColumnSplitter` constraints.

Kilo59 committed:
`SqlYearMonthSplitter.param_names` values can only be 'year` or 'month'?

Kilo59 committed:
test catching bad configs

Kilo59 committed:
extra test case

Kilo59 committed:
Merge branch 'develop' into f/1448/splitters-sede

Kilo59 committed:
fix `param_name`mypy invariant issue

Kilo59 committed:
Merge remote-tracking branch 'origin/f/1448/splitters-sede' into f/1448/splitters-sede

Kilo59 commented on great_expectations/experimental/datasources/postgres_datasource.py:
```diff
@@ -36,16 +36,18 @@ class BatchRequestError(Exception):
     column_name: str
     method_name: str
-    param_names: List[str]
+    param_names: Sequence[str]
```
>error: Incompatible types in assignment (expression has type "List[Literal['year', 'month']]", base class "ColumnSplitter" defined the type as "List[str]")  [assignment]

>note: "List" is invariant -- see https://mypy.readthedocs.io/en/stable/common_issues.html#variance
>note: Consider using "Sequence" instead, which is covariant

billdirks replied:
This was originally a Sequence but I ran into some issues. Looks like they are resolved now if this works. That's good.

Kilo59 replied:
pydantic will serialize it as a list.
https://docs.pydantic.dev/usage/types/#typing-iterables

But I think mypy will complain if you mutate a sequence, maybe that was the issue 
🤷

Kilo59 commented on great_expectations/experimental/datasources/postgres_datasource.py:
```diff
@@ -36,16 +36,18 @@ class BatchRequestError(Exception):
 class SqlYearMonthSplitter(ColumnSplitter):
-    method_name: str = "split_on_year_and_month"
-    param_names: List[str] = pydantic.Field(default_factory=lambda: ["year", "month"])
+    method_name: Literal["split_on_year_and_month"] = "split_on_year_and_month"
+    param_names: List[Literal["year", "month"]] = pydantic.Field(
```
The use of `Literal`s here means we will get nice pydantic errors if there's a typo in the config.
https://docs.pydantic.dev/usage/types/#literal-type

billdirks replied:
I'm hoping to generalize this to a general datetime splitter and get rid of default values here. We can deal with that change later but wanted to let you know.

Kilo59 committed:
Merge branch 'develop' into f/1448/splitters-sede

NathanFarmer commented on great_expectations/experimental/datasources/postgres_datasource.py:
```diff
@@ -115,7 +117,7 @@ class TableAsset(DataAsset):
     type: Literal["table"] = "table"
     table_name: str
-    column_splitter: Optional[ColumnSplitter] = None
+    column_splitter: Optional[SqlYearMonthSplitter] = None
```
Why is this specific splitter expected? Won't there be more `ColumnSplitter`s to choose from?

Kilo59 replied:
Yes, but this is the only one that exists now.
As we add more this will become a union of all the possible concrete splitters.

NathanFarmer replied:
Will there be no contributor `ColumnSplitter`s (will the universe of `ColumnSplitter`s be finite for a `Union`)?

billdirks replied:
This is a good point @NathanFarmer. Likely splitters will be a point of extensibility.

Kilo59 replied:
@NathanFarmer I'm not sure I understand. How does having a `Union` of splitter types preclude contributor-provided splitter types?

Is the point that this wouldn't work for `ColumnSplitter` plugins? That's true. We should discuss this more.

Kilo59 replied:
We discovered that the existing splitters ARE NOT a point of extensibility and are restricted to these enums.
https://github.com/great-expectations/great_expectations/blob/c0aa33feedb23671d75f3f15408171c8ee417c3b/great_expectations/execution_engine/split_and_sample/data_splitter.py#L47-L60

 As such we are adding a custom validator to ensure that provided `method_name` matches this enum.

billdirks reviewed (commented):
This LGTM. I would address @NathanFarmer's comment before merging.

Kilo59 committed:
`_splitter_method_exists` validator

Kilo59 committed:
Merge remote-tracking branch 'origin/f/1448/splitters-sede' into f/1448/splitters-sede

Kilo59 committed:
finish validator implementation

Kilo59 committed:
`SplitterMethod.members()` stub

Kilo59 committed:
finish `SplitterMethod.members()` implementation

Kilo59 committed:
more tests

Kilo59 committed:
cleanup

Kilo59 committed:
Merge branch 'develop' into f/1448/splitters-sede

Kilo59 committed:
revert `members` method

Kilo59 committed:
Merge remote-tracking branch 'origin/f/1448/splitters-sede' into f/1448/splitters-sede

Kilo59 committed:
revert import

Kilo59 committed:
link to pydantic enum docs

github-actions[bot] committed:
Merge develop into f/1448/splitters-sede

github-actions[bot] committed:
Merge develop into f/1448/splitters-sede

github-actions[bot] committed:
Merge develop into f/1448/splitters-sede

Kilo59 committed:
Merge branch 'develop' into f/1448/splitters-sede

github-actions[bot] committed:
Merge develop into f/1448/splitters-sede

github-actions[bot] committed:
Merge develop into f/1448/splitters-sede

billdirks reviewed (approved):
LGTM, thanks!
